### PR TITLE
Nominating and checking validity for `ConsensusData` containing slashed validators

### PR DESCRIPTION
**Pull request by linked0**
This is implementation about nominating and checking the validity for `ConsensusData` containing the merkleized `preimage_root` of all the preimages and the UTXO indices of validators that are missing to relay their pre-images.

This PR does not have some requirements as follows but the missing requirements will be added in other PRs soon.
* Check for `ConsensusData` which tries to melt a UTXO which is in the MPV
* Changing getRandomSeed

Part of #1076

### CODE REVIEW

**Comment by TrustHenry** [comment]
> The `SlashPolicy` clas manages the slashing policy for the misbehaving

clas -> class
There is a typo in the second commit message.

File: source/agora/consensus/SlashPolicy.d [hunk 1]
```diff
@@ -158,14 +164,67 @@ public class SlashPolicy
     private bool checkValidValidator (in Height height, in Hash utxo_key)
         @safe nothrow
     {
+        scope(failure) assert(0);
         auto preimage = this.enroll_man.getValidatorPreimage(utxo_key);
         auto enrolled = this.enroll_man.getEnrolledHeight(preimage.enroll_key);
+
```
Comment (TrustHenry) [inline source/agora/consensus/SlashPolicy.d]: ?

File: source/agora/node/Ledger.d [hunk 2]
```diff
@@ -380,6 +385,7 @@ public class Ledger
     protected void updateUTXOSet (const ref Block block) @safe
     {
         const height = block.header.height;
+
```
Comment (TrustHenry) [inline source/agora/node/Ledger.d]: You can make a separate commit.

**Comment by TrustHenry** [comment]
I'll have to look at it more.
I think it's good. It's a good start.

File: source/agora/consensus/SlashPolicy.d (hunk 1 above)
Comment (linked0) [reply source/agora/consensus/SlashPolicy.d]: Fixed it.

File: source/agora/node/Ledger.d (hunk 2 above)
Comment (linked0) [reply source/agora/node/Ledger.d]: deleted the unnecessary line.

**Comment by linked0** [comment]
> > The `SlashPolicy` clas manages the slashing policy for the misbehaving
> 
> clas -> class
> There is a typo in the second commit message.

modified the message.

**Comment by linked0** [comment]
Need to split the last commit into smaller commits.

File: source/agora/consensus/SlashPolicy.d [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+        Params:
+            missing_validators = will contain the validators
+            height = the desired block height to look up the validators for
+
+    ***************************************************************************/
+
+    public void getMissingValidators (ref uint[] missing_validators,
+        in Height height) @safe nothrow
+    {
+        Hash[] keys;
```
Comment (AndrejMitrovic) [inline source/agora/consensus/SlashPolicy.d:100]: If you plan to use this as a buffer then I think you should add:

```
missing_validators.length = 0;
() @trusted { assumeSafeAppend(missing_validators); }();
```

Otherwise if you don't want it to be a buffer you can use `out uint[]`. But it would mean there is always an allocation. So the first approach might be better.

File: source/agora/consensus/SlashPolicy.d [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+        Params:
+            height = the desired block height to look up the hash for
+            utxo_key = the UTXO key idendifying a validator
+
+        Returns:
+            true if the validator has a right pre-image
+
+    ***************************************************************************/
+
+    private bool checkValidValidator (in Height height, in Hash utxo_key)
```
Comment (AndrejMitrovic) [inline source/agora/consensus/SlashPolicy.d]: I would name this something better like `hasRevealedPreimage`. Because the concept of a "valid validator" doesn't really sound informative.

File: source/agora/consensus/SlashPolicy.d [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+        }
+
+        Hash[] valid_keys;
+        foreach (key; keys)
+        {
+            if (checkValidValidator(height, key))
+                valid_keys ~= key;
+        }
+
+        return this.enroll_man.getRandomSeed(valid_keys, height);
```
Comment (AndrejMitrovic) [inline source/agora/consensus/SlashPolicy.d:151]: That's not a merkle root though.

If you want we can just rename it to something like `random_seed` in the `ConsensusData` and in the `Block`. I think it's fine to just store the value of `getRandomSeed()`, it's just that the naming is wrong because that function doesn't actually create a merkle root.

File: source/agora/consensus/SlashPolicy.d [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+    }
+
+    /***************************************************************************
+
+        Check if the the validator is valid for the height
+
+        This checks if a validator has a pre-image for the height
```
Comment (AndrejMitrovic) [inline source/agora/consensus/SlashPolicy.d]: You can remove the first sentence, the second one already describes what the function does.

File: source/agora/consensus/SlashPolicy.d [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+
+        Params:
+            height = the desired block height to look up the hash for
+            utxo_key = the UTXO key idendifying a validator
+
+        Returns:
+            true if the validator has a right pre-image
```
Comment (AndrejMitrovic) [inline source/agora/consensus/SlashPolicy.d]: > has a right pre-image

has revealed its preimage for the provided block height

File: source/agora/consensus/SlashPolicy.d [hunk 8]
```diff
@@ -40,7 +40,8 @@ import agora.consensus.state.UTXODB;
 import agora.utils.Log;
```
Comment (AndrejMitrovic) [inline source/agora/consensus/SlashPolicy.d:40]: The comment in the commit should be changed, because the preimage root is not merkleized. It's just a hash concatenation of all the preimages which were sorted by the utxo key.

File: source/agora/consensus/protocol/Nominator.d [hunk 9]
```diff
@@ -544,6 +553,21 @@ extern(D):
...
+                if (auto fail_reason = this.ledger.validateSlashingData(data))
+                {
+                    log.error("validateValue(): Preimage Validation failed: {}. Data: {}",
+                        fail_reason, data);
+                    return ValidationLevel.kInvalidValue;
+                }
+            }
+            else
+            {
+                temp_level = ValidationLevel.kMaybeValidValue;
```
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d]: `return ValidationLevel.kMaybeValidValue`

File: source/agora/consensus/protocol/Nominator.d [hunk 10]
```diff
@@ -534,6 +542,7 @@ extern(D):
     {
         const cur_time = this.clock.networkTime();
         const exp_time = this.getExpectedBlockTime(Height(slot_idx));
+        ValidationLevel temp_level;
```
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d]: We don't need a state variable for this, we can use early returns. I'll add more comments below.

File: source/agora/consensus/protocol/Nominator.d [hunk 11]
```diff
@@ -544,6 +553,21 @@ extern(D):
...
+
+            bool is_same = this.ledger.isSameMissingValidatorsAsLocal(data);
+            if (is_same)
+            {
+                if (auto fail_reason = this.ledger.validateSlashingData(data))
+                {
+                    log.error("validateValue(): Preimage Validation failed: {}. Data: {}",
+                        fail_reason, data);
+                    return ValidationLevel.kInvalidValue;
+                }
```
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d]: Add `return ValidationLevel. kFullyValidatedValue` here and then you won't need the `temp_level` anymore.

File: source/agora/node/Ledger.d [hunk 12]
```diff
@@ -494,6 +506,56 @@ public class Ledger
...
+        Params:
+            data = consensus data
+
+        Returns:
+            the error message if validation failed, otherwise null
+
+    ***************************************************************************/
+
+    public string validateSlashingData (ConsensusData data) @trusted
```
Comment (AndrejMitrovic) [inline source/agora/node/Ledger.d]: `in ConsensusData data`

Comment (AndrejMitrovic) [reply source/agora/node/Ledger.d]: Can we make this function `@safe` and also mark `isInvalidPreimageRootReason` as `@safe` too if needed? We should avoid labeling everything as `@trusted`, because it tells the compiler to avoid doing any kind of safety checks.

File: source/agora/consensus/protocol/Nominator.d [hunk 13]
```diff
@@ -544,6 +553,21 @@ extern(D):
                     fail_reason, data);
                 return ValidationLevel.kInvalidValue;
             }
+
+            bool is_same = this.ledger.isSameMissingValidatorsAsLocal(data);
+            if (is_same)
+            {
+                if (auto fail_reason = this.ledger.validateSlashingData(data))
```
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d]: `isSameMissingValidatorsAsLocal` and `validateSlashingData` (which forwards to `isInvalidPreimageRootReason`) seem to have similar checks. They both check the `missing_validators` array.

Actually thinking about this, maybe we can just remove `preimage_root` from the `ConsensusData`? The root can be easily calculated after consensus is reached, when the block is about to be created and signed. So we wouldn't have to check its validity here, we would just have to compare the two arrays. What do you think, is it a good idea?

In fact I think we only need to verify the root when downloading blocks from other nodes.

**Review by AndrejMitrovic — CHANGES_REQUESTED** [review]
Completed first round of review. Will wait for some comments.

**Comment by AndrejMitrovic** [comment]
Great work so far @linked0!

File: source/agora/consensus/protocol/Nominator.d (hunk 13 above)
Comment (linked0) [reply source/agora/consensus/protocol/Nominator.d]: > maybe we can just remove preimage_root from the ConsensusData?

I think we can do it because all the nodes have the same enrolled UTXOs (a.k.a. Validators) and we can calculate the root only with the indices. It's a good idea.

> In fact I think we only need to verify the root when downloading blocks from other nodes.

In fact, how to process the catch-up is a blocker for me now. There is some problem in checking the validity of the root in the middle of downloading blocks from other nodes. Suppose that the latest height is 1009 and all the current validators have enrolled at the height of 1008. When downloading the blocks between 0 and 1008, we can not check the validity of the root because there are no pre-images for un-enrolled validators.

File: source/agora/node/Ledger.d (hunk 12 above)
Comment (linked0) [reply source/agora/node/Ledger.d]: > Can we make this function @safe and also mark isInvalidPreimageRootReason as @safe too if needed?

We can. I'm not gonna use `@trust` as far as possible.

File: source/agora/consensus/protocol/Nominator.d (hunk 13 above)
Comment (AndrejMitrovic) [reply source/agora/consensus/protocol/Nominator.d]: > there are no pre-images for un-enrolled validators.

That's why we should keep the last preimage and enrollment of every historic validator. It should be possible if we follow my new suggestion here: https://github.com/bpfkorea/agora/issues/1297#issuecomment-742203550

Then full nodes which want to validate really old blocks where those validators don't exist anymore would just use the `getPreimage` API to find those outdated validator's last preimages.

The problem right now is that we delete items from the `validator_set` as soon as a node gets un-enrolled. But we should probably keep that data forever so historic blocks can be validated. Of course, you would still need some information about which validators are **inactive**. Currently they're marked inactive by removing them from the validator set database, but maybe we should just add a status flag instead?

Comment (linked0) [reply source/agora/consensus/protocol/Nominator.d]: > Currently they're marked inactive by removing them from the validator set database, but maybe we should just add a status flag instead?

The state could solve my problem. But we should think about the situation that the data of `validator_set` will be ever-increasing, which might not be so big though.

**Comment by linked0** [comment]
Ready for review after applying the reviewers' suggestions, the flaky failures on the CI builds keep happening though.
I'll keep checking it.

**Comment by AndrejMitrovic** [comment]
> the flaky failures on the CI builds keep happening though.

> /home/runner/work/agora/agora/submodules/ocean/src/ocean/text/convert/Formatter.d:466: error: undefined reference to '_D6geod247bitblob__T7BitBlobVmi512ZQp8toStringMxFNfMDFNfAxaZvZv'

It seems it's some kind of linker error.

**Comment by linked0** [comment]
> > the flaky failures on the CI builds keep happening though.
> 
> > /home/runner/work/agora/agora/submodules/ocean/src/ocean/text/convert/Formatter.d:466: error: undefined reference to '_D6geod247bitblob__T7BitBlobVmi512ZQp8toStringMxFNfMDFNfAxaZvZv'
> 
> It seems it's some kind of linker error.

Can I get any clue why the CI build on my PR keeps failing while the build on Chris's is going well now? The build on my local machine is going well.

**Comment by linked0** [comment]
I missed rebasing it. After rebasing it, I'm trying to ci build again.
I found with Chris's advice that it occurs only on the command of "dub build -c client --skip-registry=all --compiler=ldc2".

**Comment by AndrejMitrovic** [comment]
Maybe you need to add more modules to https://github.com/bpfkorea/agora/blob/ea9f26c3d09527f79c16365633923b472393571e/dub.json#L96?

**Comment by linked0** [comment]
> Maybe you need to add more modules to
> 
> https://github.com/bpfkorea/agora/blob/ea9f26c3d09527f79c16365633923b472393571e/dub.json#L96
> 
> ?

will check it, thanks.

**Comment by linked0** [comment]
The ci builds are going well.
The position of the "version (unittest):" line in a network test file was wrong.

File: source/agora/consensus/protocol/Data.d [hunk 14]
```diff
@@ -13,6 +13,7 @@
 
 module agora.consensus.protocol.Data;
 
+import agora.common.Types;
```
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Data.d]: Unnecessary import?

File: source/agora/consensus/SlashPolicy.d [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+        Hash[] keys;
+        if (!this.enroll_man.getEnrolledUTXOs(keys) || keys.length == 0)
+        {
+            log.fatal("Could not retrieve enrollments / no enrollments found");
+            assert(0);
+        }
+
+        foreach (idx, utxo_key; keys)
+        {
+            if (!hasRevealedPreimage(height, utxo_key))
```
Comment (AndrejMitrovic) [inline source/agora/consensus/SlashPolicy.d]: `this.hasRevealedPreimage`

File: source/agora/consensus/SlashPolicy.d [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+        if (!this.enroll_man.getEnrolledUTXOs(keys) || keys.length == 0)
+        {
+            log.fatal("Could not retrieve enrollments / no enrollments found");
+            assert(0);
+        }
+
+        Hash[] valid_keys;
+        foreach (key; keys)
+        {
+            if (hasRevealedPreimage(height, key))
```
Comment (AndrejMitrovic) [inline source/agora/consensus/SlashPolicy.d]: `this.hasRevealedPreimage`

File: source/agora/consensus/SlashPolicy.d [hunk 17]
```diff
@@ -141,6 +142,11 @@ public class SlashPolicy
                 valid_keys ~= key;
         }
 
+        // NOTE: The preimage root of `Hash.init` value is currently
+        // checked in the `validateSlashingData` function of `Ledger`.
+        if (valid_keys.length == 0)
+            return Hash.init;
```
Comment (AndrejMitrovic) [inline source/agora/consensus/SlashPolicy.d:149]: This should be documented in the `Returns` section.

File: source/agora/consensus/SlashPolicy.d [hunk 18]
```diff
@@ -169,6 +175,48 @@ public class SlashPolicy
...
+        const ref uint[] missing_validators) @safe
+    {
+        Hash[] keys;
+        if (!this.enroll_man.getEnrolledUTXOs(keys) || keys.length == 0)
+        {
+            log.fatal("Could not retrieve enrollments / no enrollments found");
+            assert(0);
+        }
+
+        Hash[] valid_keys;
```
Comment (AndrejMitrovic) [inline source/agora/consensus/SlashPolicy.d]: This doesn't seem to be used for anything other than appending to it. I think it should be removed.

File: source/agora/consensus/SlashPolicy.d [hunk 19]
```diff
@@ -169,6 +175,48 @@ public class SlashPolicy
...
+        uint[] local_missing_validators;
+        foreach (idx, key; keys)
+        {
+            if (!hasRevealedPreimage(height, key))
+                local_missing_validators ~= cast(uint)idx;
+
+            if (missing_validators.find(idx).empty())
+                valid_keys ~= key;
```
Comment (AndrejMitrovic) [inline source/agora/consensus/SlashPolicy.d]: Ditto, this should be removed.

File: source/agora/consensus/SlashPolicy.d [hunk 20]
```diff
@@ -169,6 +175,48 @@ public class SlashPolicy
...
+        {
+            log.fatal("Could not retrieve enrollments / no enrollments found");
+            assert(0);
+        }
+
+        Hash[] valid_keys;
+        uint[] local_missing_validators;
+        foreach (idx, key; keys)
+        {
+            if (!hasRevealedPreimage(height, key))
```
Comment (AndrejMitrovic) [inline source/agora/consensus/SlashPolicy.d]: `this.hasRevealedPreimage`

File: source/agora/consensus/SlashPolicy.d [hunk 21]
```diff
@@ -169,6 +175,48 @@ public class SlashPolicy
...
+            missing_validators = list of indices to the validator UTXO set
+                which have not revealed the preimage
+
+        Returns:
+            `null` if the information is valid at the proposed height,
+            otherwise a string explaining the reason it is invalid.
+
+    ***************************************************************************/
+
+    public string isInvalidPreimageRootReason (Height height,
```
Comment (AndrejMitrovic) [inline source/agora/consensus/SlashPolicy.d:195]: In the future it would be more efficient to construct the local array **once**, each time a new preimage is received. That way you can cache it and then the comparison is simple:

```
remote_missing == local_missing
```

**Review by AndrejMitrovic — CHANGES_REQUESTED** [review]
Just a few comments.

File: source/agora/consensus/SlashPolicy.d (hunk 21 above)
Comment (linked0) [reply source/agora/consensus/SlashPolicy.d:195]: > In the future it would be more efficient to construct the local array once, each time a new preimage is received. That way you can cache it and then the comparison is simple:
> 
> remote_missing == local_missing

Sure, I will. We had a similar requirement for the `preimage_root` which is not used though. We can apply the same concept for the variable for the `MPV`.

**Comment by linked0** [comment]
Ready for review.
- removed unnecessary import and variable
- added more explanation for the function description
- applied requests about the code

File: source/agora/consensus/protocol/Nominator.d [hunk 22]
```diff
@@ -544,6 +552,14 @@ extern(D):
                     fail_reason, data);
                 return ValidationLevel.kInvalidValue;
             }
+
```
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d:555]: You could do one more check here. If the array is bigger than the total number of active validators, then it's clearly invalid. So in that case you can return `kInvalidValue`.

**Review by AndrejMitrovic — APPROVED** [review]

**Comment by AndrejMitrovic** [comment]
There are some possible improvements, but I don't want to delay this any longer so I'm approving it.

File: source/agora/consensus/protocol/Nominator.d (hunk 22 above)
Comment (linked0) [reply source/agora/consensus/protocol/Nominator.d:555]: > You could do one more check here. If the array is bigger than the total number of active validators, then it's clearly invalid. So in that case you can return kInvalidValue.

Ah, Right. It should return `kInvalidValue` in that case. I will fix it with the other requirements you suggested.
Thank you.

**Comment by AndrejMitrovic** [comment]
@linked0 I'll merge this so we can move forward with it!
